### ISSUE TITLE
Clients with multiple types and/or security contexts

### ISSUE THREAD

**Issue by DruIX** [bshaffer/oauth2-server-php#550]
First of all, thanks to @bshaffer and the rest of the community that's put work into this library. I've used it on a number of occasions and it's saved a huge amount of time.

I'm now working on an application that will be part of a setup similar to the one mentioned in #257. There is a separate Auth server (using this lib, AS), an API (Resource Server, RS) and a JavaScript web app that makes use of that API. 

This creates a bit of a problem as the RS and the JS app are really both part of the same client (it's a single page application), but the JS app is considered public whereas the RS is considered confidential. When using the "password" grant type for login, the JS app would be authenticating as a public client, but later on if the RS needs to make requests to the AS, authenticating using "client_credentials", it would need a client secret to do so.

The core of the problem is in the fact that the AS and RS are separated. The RS needs to be able to examine the token attached to requests from the JS app and check that they are valid for that particular client. Whether this is done via JSON Web Tokens or an implementation of ["Introspection"](https://tools.ietf.org/html/draft-ietf-oauth-introspection-06), the recipient client of the original token needs to match the client_id of the RS.

This structure is mentioned in [RFC 6749 Section 2.1](http://tools.ietf.org/html/rfc6749#section-2.1):
> A client may be implemented as a distributed set of components, each with a different client type and security context (e.g., a distributed client with both a confidential server-based component and a public browser-based component).  If the authorization server does not provide support for such clients or does not provide guidance with regard to their registration, the client SHOULD register each component as a separate client.

Can this be worked in at all? Or have I simply misunderstood something? Although the RFC states that I should fall back to using separate clients for each component, it doesn't feel like a particularly robust solution going forward.

**Comment by DruIX** [comment]
Whoops! It looks like I missed #277, where most of this was answered.

I still think that allowing a single client to be public and confidential in certain circumstances would be useful and would help with streamlining things (especially when it comes to separating the AS from the RS)

Any thoughts?

**Comment by bshaffer** [comment]
Yes, my thoughts had been it is best to make clients specific to the grant type you are using. I suppose a problem with this is when a user authorizes via `Implicit` and the server wishes to make backend-calls using `Authorization Code`. Having two authentication flows for two different clients would be very strange.

In this case, you could override the `OAuthClient` storage, and make `isPublicClient` be based off a database flag rather than the secret. But... since the spec clearly allows for clients to be used for both flows, I agree this would make a good enhancement.
